부산대 BE_오선재 3단계 - DB 연결

5seonjae opened the pull request:
## ✨ 3단계: DB 연동 (JdbcTemplate 기반 CRUD 구현)

메모리 기반 CRUD에서 DB 연동으로 전환하는 과정에서,
초기에는 각 기능(`save`, `findAll`, `update`, `delete`)마다 커밋을 나누려 했습니다.

하지만 메모리 코드를 제거하고 DB 기반 코드로 교체하는 과정에서
기능 하나만 따로 테스트하기가 애매했고, 연쇄적으로 수정이 필요한 구조라
결국 전체를 한 번에 수정하고 통합적으로 테스트한 뒤 커밋했습니다.

혹시 이런 경우에도 기능 단위로 쪼개서 커밋하고 테스트할 수 있는 방법이 있을지 궁금합니다.

현재는 H2 메모리 DB를 사용했으며, 운영 환경에서는 MySQL과 같은 RDBMS로의 전환이 필요할 것 같습니다.

---

### ✅ 구현 내용

- `application.properties`에 H2 설정 추가
- `schema.sql`, `data.sql`을 통한 자동 테이블 생성 및 샘플 데이터 삽입
- `JdbcTemplate` 기반 CRUD 구현
- 기존 메모리 저장소 완전히 제거
- 상품 등록, 조회, 수정, 삭제 기능 정상 동작 확인

---

### 🧪 테스트 방법

- 서버 실행 후 `/admin/products`에서 전체 기능 테스트 완료
- H2 Console에서 데이터 확인 (`jdbc:h2:mem:testdb`)

5seonjae committed:
refactor: 상품 등록 시 Product 생성 방식을 생성자 주입으로 변경

- ProductViewController에서 setter 방식 대신 생성자 방식으로 Product 객체 생성
- 도메인 객체의 불변성과 무결성 보장을 위한 리팩토링
- Product에서 setter 제거

5seonjae committed:
docs: 3단계(DB 연동) 구현 목표 및 디렉터리 구조 README에 추가

- H2 데이터베이스 연동 및 JdbcTemplate 기반 CRUD 구현 목표 설명 추가
- 메모리 기반 저장소 제거 및 SQL 스크립트 활용 항목 정리
- 프로젝트 디렉터리 구조 최신화

5seonjae committed:
docs: 3단계 기능 구현 체크리스트 상태 수정

- 아직 구현되지 않은 기능에 대해 체크 표시 제거
- README.md의 구현 진행 상황을 실제 상태와 일치하도록 수정

5seonjae committed:
fix: setter 대신 새 객체를 생성하여 상품 수정 시 ID 유지 방식 적용

- setName 등 setter 사용 대신 생성자 기반으로 Product 객체 재생성
- 기존 ID는 setId로 유지하고, 수정 후 storage에 다시 반영
- 객체 불변성 유지와 책임 분리를 위한 리팩토링

5seonjae committed:
chore: H2 데이터베이스 연동 및 SQL 자동 실행 설정 추가

- spring.datasource.url, username 설정으로 H2 메모리 DB 연결
- spring.sql.init.mode=always 설정으로 schema.sql 자동 실행
- spring.h2.console 설정 추가로 콘솔 접속 가능하도록 구성

5seonjae committed:
chore: application.properties 인코딩을 UTF-8로 변경하고 한글 주석 복원

5seonjae committed:
feat: products 테이블 생성 SQL 추가

5seonjae committed:
feat: 초기 샘플 상품 데이터를 위한 data.sql 추가

5seonjae committed:
refactor: 기존 메모리 저장소 제거 및 JDBC 기반 상품 CRUD 전환 마무리

5seonjae committed:
docs: 3단계 구현 완료 체크리스트 업데이트

iphwade commented on README.md:
```diff
@@ -183,11 +183,11 @@ JdbcTemplate을 활용해 상품의 등록, 조회, 수정, 삭제 기능을 데
+- [ ] `data.sql`로 초기 상품 데이터 삽입
+- [ ] `application.properties`에 H2 설정 추가
+- [ ] `JdbcTemplate`을 통한 상품 등록/조회/삭제 구현
+- [ ] 메모리 저장소 코드 제거
```
첫 커밋인데 x 처리가 되어있어 의아했었는데 바로 수정해주셨군요ㅋㅋㅋㅋ

iphwade commented on src/main/resources/application.properties:
```diff
@@ -1 +1,13 @@
 spring.application.name=spring-gift
+
+# H2 DB ??
```
물음표는 어떤 의미일까요-?

iphwade commented on src/main/resources/application.properties:
```diff
@@ -1,13 +1,13 @@
 spring.datasource.password=
 
-# SQL ???? ?? ??
+# SQL 스크립트 자동 실행 설정
```
헛 바로 다음 커밋에서 수정이 되었군요!

iphwade commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE products
+(
+    id BIGINT PRIMARY KEY AUTO_INCREMENT,
+    name VARCHAR(255) NOT NULL,
```
이름은 255보다 더 작게 설정할 수 있을까요?

5seonjae replied:
> 이름은 255보다 더 작게 설정할 수 있을까요? 

> URL은 255를 넘을 수도 있을까요? 

좋은 지적 감사합니다! 말씀해주신 대로 두 필드 모두 길이 제한을 다시 생각해볼 필요가 있을 것 같습니다.

name의 경우, 실제로는 100자 이내로 충분할 것 같아 VARCHAR(100) 정도로 줄여보는 것도 고려하고 있습니다.

image_url은 말씀 주신 대로 이미지 호스팅 서비스에 따라 255자를 초과할 가능성도 있어서, TEXT나 VARCHAR(512) 정도로 늘리는 것도 대안이 될 수 있을 것 같습니다.


현재는 우선 VARCHAR(255)로 통일했지만, 실사용 데이터를 고려해서 길이 제한을 조정해보겠습니다!

iphwade replied:
https://github.com/next-step/spring-gift-product/pull/192#discussion_r2174331845
네-! 위 내용도 참고차 읽어보시면 도움될 것 같아요-!

iphwade commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,7 @@
+    id BIGINT PRIMARY KEY AUTO_INCREMENT,
+    name VARCHAR(255) NOT NULL,
+    price INT NOT NULL,
+    image_url VARCHAR(255) NOT NULL
```
URL은 255를 넘을 수도 있을까요?

iphwade commented on src/main/resources/data.sql:
과자점인가보군요!

iphwade commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -13,36 +16,52 @@
+            rs.getInt("price"),
+            rs.getString("image_url")
+        );
+        product.setId(rs.getInt("id"));
```
id는 따로 설정하신 이유가 있으실까요-?

5seonjae replied:
처음에는 id 값을 ResultSet에서 따로 주입해줘야 하다 보니 setId() 메서드를 사용했었는데요,
프로젝트 전반적으로 Product를 불변 객체(Immutable Object)로 설계하려는 방향과 맞지 않아
setId() 메서드를 제거하고, 생성자에서 id를 포함해 초기화하는 방식으로 수정했습니다.

즉, 등록 시점에는 id가 null일 수 있으므로 Long 타입으로 선언하고,
조회/수정 시에는 id가 있는 생성자를 사용하여 객체를 생성하는 구조로 바꿨습니다 :)

iphwade replied:
불변 객체에 대해서도 알고계셨군요-! :)
상황에 따라 불변 객체가 불리한 경우도 있지만 대부분의 경우 불변객체 이점이 더욱 많습니다.

iphwade commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -13,34 +16,52 @@
-            existing.setPrice(dto.getPrice());
-            existing.setImageUrl(dto.getImageUrl());
-        }
+    public void update(Long id, ProductUpdateRequestDto productUpdateRequestDto) {
```
DTO가 Repository 레이어까지 전달되는군요.
뷰 레이어의 변경이 Repository 레이어까지 영향을 미칠 수 있는 구조로 보이는데요.
분리하는 게 좋아보입니다.

5seonjae replied:
좋은 피드백 감사합니다!
말씀해주신 대로 DTO가 Repository까지 전달되면
뷰(View)나 외부 요청의 변화가 곧바로 Repository 레이어에 영향을 줄 수 있어
레이어 간 결합도가 높아지는 문제가 발생할 수 있겠다는 점을 인지하게 되었습니다.

이 부분은 ProductUpdateRequestDto를 Repository에 넘기지 않고,
Service 레이어에서 도메인 객체로 변환한 후 Repository로 넘기는 방식으로
책임을 명확히 분리하도록 수정해보겠습니다!

iphwade replied:
좋은 방향으로 잘 수정되었군요-! :)

iphwade reviewed (changes requested):
안녕하세요. 선재님.
이번에도 잘 구현해주셨습니다. 
몇가지 코멘트 남겼는데요~ 읽고 답변주시면 리뷰하는데 도움될 것 같아요~

iphwade commented:
> ## ✨ 3단계: DB 연동 (JdbcTemplate 기반 CRUD 구현)
> 메모리 기반 CRUD에서 DB 연동으로 전환하는 과정에서, 초기에는 각 기능(`save`, `findAll`, `update`, `delete`)마다 커밋을 나누려 했습니다.
> 
> 하지만 메모리 코드를 제거하고 DB 기반 코드로 교체하는 과정에서 기능 하나만 따로 테스트하기가 애매했고, 연쇄적으로 수정이 필요한 구조라 결국 전체를 한 번에 수정하고 통합적으로 테스트한 뒤 커밋했습니다.
> 
> 혹시 이런 경우에도 기능 단위로 쪼개서 커밋하고 테스트할 수 있는 방법이 있을지 궁금합니다.

커밋은 분리하되 테스트는 한번에 하는 방법이 좋을 것 같네요-! 
아직은 요구사항에 없었지만 테스트 코드를 작성해보면 분리하는데 도움이 되겠네요.

iphwade commented:
지금도 정말 잘해주셨는데요~ DB가 변경되는 상황에 기존 ProductRepository는 남겨두고 JdbcProductRepository를 새롭게 작성하는 형태로 접근해보면 Spring에서 Bean을 관리해주는 장점과 Java의 Interface를 사용하는 것이 주는 이점을 쉽게 확인해보실 수 있으셨을 것 같아요.

5seonjae commented:
> 커밋은 분리하되 테스트는 한번에 하는 방법이 좋을 것 같네요-! 
> 아직은 요구사항에 없었지만 테스트 코드를 작성해보면 분리하는데 도움이 되겠네요. 
> 

커밋은 따로 하되 테스트는 한 번에 하는 방법이 있었군요!
저는 커밋을 할 때마다 해당 코드들이 정상적으로 실행되어야 한다고 생각해서, 결국 모든 기능을 한 번에 수정하고 통합 테스트 후 커밋하는 방향으로 진행했던 것 같습니다.

다음에는 각 기능 구현이 끝날 때마다 커밋을 하고, 테스트는 한 번에 통합적으로 진행하는 방식으로 시도해보겠습니다.

추가로, 이런 방식도 고려해볼 수 있을 것 같습니다:

1. 기존 ProductRepository는 유지하고, JdbcProductRepository를 새로 만들어서 각 기능 단위로 구현 및 테스트해보는 방법


2. 이번 미션에서는 브랜치를 나누는 것이 어렵긴 했지만, 가능하다면 기능별 브랜치를 만들어 작업하고, 각 브랜치에서 커밋 및 테스트 후 병합하는 방식



이런 방법들을 통해 다음 단계에서는 기능 단위 커밋과 테스트의 균형을 더 잘 맞춰볼 수 있을 것 같습니다.

그리고, 간단한 테스트 코드도 작성해보면서 구조나 커밋 단위를 나누는 기준에 대해 연습해보겠습니다.

5seonjae commented:
💬 추가 질문 드립니다!
기능을 수정할 때, 해당 기능이 여러 메서드로 나뉘어 있어서 각각의 메서드를 수정해야 하는 상황이라면,
하나의 기능 단위로 커밋하는 것이 좋을지, 아니면 메서드 단위로 나누어 커밋하는 것이 더 나은지 궁금합니다.

예를 들어 save() 기능을 수정하면서,

Repository

Service

Controller
세 레이어에서 각각 메서드를 수정하게 되는 경우, 기능 단위로 하나의 커밋을 만드는 게 더 적절할까요?
아니면 각 메서드(혹은 레이어) 단위로 나누는 게 더 명확할까요?


이런 상황에서 커밋 단위를 나누는 기준이나 팁이 있다면 조언 부탁드립니다!

5seonjae commented:
> 지금도 정말 잘해주셨는데요~ DB가 변경되는 상황에 기존 ProductRepository는 남겨두고 JdbcProductRepository를 새롭게 작성하는 형태로 접근해보면 Spring에서 Bean을 관리해주는 장점과 Java의 Interface를 사용하는 것이 주는 이점을 쉽게 확인해보실 수 있으셨을 것 같아요.

감사합니다!
말씀해주신 방식대로 ProductRepository는 인터페이스로 두고,
JdbcProductRepository를 새로 구현하는 식으로 분리했으면
Spring의 Bean 관리나 DI 구조를 좀 더 명확히 경험해볼 수 있었겠다는 생각이 듭니다.

이번엔 구조를 한 번에 바꾸면서 실감하지 못했지만,
다음 미션에서는 이 구조를 적용해보고, 테스트 코드까지 함께 작성하면서
Spring의 DI와 인터페이스 분리의 이점을 직접 체험해보겠습니다!

5seonjae committed:
refactor: Product 엔티티 불변 객체로 리팩토링 및 생성자 분리

- DB 자동 생성 id에 대응하기 위해 id 필드 타입을 long → Long으로 변경
- 생성 시점에만 모든 필드를 설정하도록 생성자 설계
- 등록용(id 제외)과 조회/수정용(id 포함) 생성자 분리로 명확한 책임 부여

5seonjae committed:
refactor: Repository에서 DTO 의존 제거 및 Product 기반 update 메서드 리팩토링

- ProductRepository.update() 메서드가 더 이상 ProductUpdateRequestDto를 직접 받지 않도록 수정
- 서비스 레이어에서 DTO를 도메인 객체(Product)로 변환한 뒤 Repository에 전달
- 계층 간 의존성을 줄이고, 책임을 명확히 분리함

5seonjae committed:
refactor: updateProduct 로직에서 불변 객체 생성 방식으로 리팩토링

- Product가 setter가 없는 불변 객체이므로 수정 불가능한 기존 인스턴스 대신 새로운 인스턴스를 생성하여 업데이트
- Repository에 DTO가 전달되지 않도록 도메인 객체로 변환 후 전달
- 계층 간 역할 분리 및 객체 불변성 보장

iphwade commented:
> 💬 추가 질문 드립니다! 기능을 수정할 때, 해당 기능이 여러 메서드로 나뉘어 있어서 각각의 메서드를 수정해야 하는 상황이라면, 하나의 기능 단위로 커밋하는 것이 좋을지, 아니면 메서드 단위로 나누어 커밋하는 것이 더 나은지 궁금합니다.
> 
> 예를 들어 save() 기능을 수정하면서,
> 
> Repository
> 
> Service
> 
> Controller 세 레이어에서 각각 메서드를 수정하게 되는 경우, 기능 단위로 하나의 커밋을 만드는 게 더 적절할까요? 아니면 각 메서드(혹은 레이어) 단위로 나누는 게 더 명확할까요?
> 
> 이런 상황에서 커밋 단위를 나누는 기준이나 팁이 있다면 조언 부탁드립니다!

사실 저는 기능에 이상이 없고 빌드가 가능한다면 최대한 잘게 나누는게 낫다고 생각합니다. 많은 수의 커밋이 나와도 리뷰할 때는 오히려 의도가 더 잘 보여져서요~ 다만, 실제로 로직이 변경되는 것이 아닌 변수의 이름을 변경한다거나 이런 사소한 변경들은 하나의 커밋에 몰아두어도 충분하다고 생각합니다~ 

앞으로 미션하시면서 더욱 고민해볼 포인트가 될 것이라고 생각해요-! 
요지는 도메인 이해도가 낮은 리뷰어가 커밋 단위로 리뷰할 때 이해하기 편할까를 기준으로 둬보면 어떨까요-?

iphwade commented on src/main/java/gift/entity/Product.java:
```diff
@@ -5,21 +5,34 @@
 
+    // 등록 시 사용하는 생성자
+    public Product(String name, int price, String imageUrl) {
+        this(null, name, price, imageUrl);  // 위 생성자 재사용
```
생성자 체이닝을 사용하셨군요-! :)

iphwade commented on src/main/java/gift/repository/ProductRepository.java:
엇? update 메서드의 메서드 시그니쳐의 변경이 있었는데 사용하는 쪽(ex, service)의 변경이 없군요?
이 커밋으로 롤백하는 상황에서 빌드가 되지 않는 상황이 발생할 수도 있을 것 같아요~ 
되도록 커밋의 단위는 빌드가 가능한 지를 기준으로 해볼까요?

5seonjae replied:
좋은 지적 감사합니다!

해당 `update` 메서드는 기존에 `id`와 분리된 형태였는데, 최근 리팩토링하면서 `Product` 하나로 통합되도록 수정했습니다.  
현재는 Service 레이어에서도 해당 변경에 맞춰 함께 수정되어 빌드 가능한 상태로 반영되어 있습니다.

말씀해주신 것처럼 커밋 단위에서도 항상 빌드가 가능한 상태를 유지하는 것이 중요하다는 점 공감합니다.  
앞으로도 변경이 생길 경우 커밋 시점에 빌드 여부를 더 신경 쓰도록 하겠습니다!

iphwade reviewed (approved):
안녕하세요~ 선재님 
하루하루 미션하시면서 생각보다 고되실텐데 잘 해내고 계신 것 같아 응원하게 됩니다~
잘 해주고 계십니다~ 

생각해볼만한 몇가지 코멘트 남겨두었습니다~

그럼 다음 미션도 화이팅입니다-! 👍
